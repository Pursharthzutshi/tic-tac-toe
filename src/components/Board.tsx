import React from "react";
import { useAppDispatch, useAppSelector } from "../hooks";
import { handleOnClick, restartGame } from "./slicers/BoardSlicer";
import "../components/board.css"
import { BoardRow } from "./BoardRow";


export const Board = () => {

    //use selectors 
    const boardState = useAppSelector(((state)=>state.BoardSlicer.board))

    // const player = useAppSelector(((state)=>state.BoardSlicer.player))
    const winner = useAppSelector(((state)=>state.BoardSlicer.winner))
    const isNotWinner = useAppSelector(((state)=>state.BoardSlicer.isNotWinner))

    // dispatch
    const dispatch = useAppDispatch();


    const sendDisptachValues = (indexArgs:any)=>{
        dispatch(handleOnClick(indexArgs))
    }

    return (
        <div className="board">

            <h3>AI Tic Tac Toe Board</h3>
            {
                boardState.map((row:any, rowIndex) => {
                    return (

                            <BoardRow row={row} rowIndex={rowIndex}/>
                    )

                })
            }

             {winner && <p>{winner === "X" ? "You Win" : "AI Wins"}</p>}
 			{isNotWinner && <p> No one wins</p>}

             <button className="reset-button" type="button" onClick={()=>dispatch(restartGame("restart"))}>Restart Button</button>


        </div>
    );
};