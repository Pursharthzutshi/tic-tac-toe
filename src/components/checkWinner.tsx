type BoardArray = Array<Array<string|null>>

export function checkWinner  (board:BoardArray): string | null {
    const lines = [
        //Rows
        [board[0][0], board[0][1], board[0][2]],
        [board[1][0], board[1][1], board[1][2]],
        [board[2][0], board[2][1], board[2][2]],
        //columns
        [board[0][0], board[1][0], board[2][0]],
        [board[0][1], board[1][1], board[2][1]],
        [board[0][2], board[1][2], board[2][2]],

        //Diagonals
        [board[0][0], board[1][1], board[2][2]],
        [board[0][2], board[1][1], board[2][0]],
    ];
  for (const line of lines) {
        const [a, b, c] = line;
        if (a && a === b && a === c) {
            return a; // Return the winner ('X' or 'O')
        }
    }

    // If no winning lines found, return null
    return null;
}
