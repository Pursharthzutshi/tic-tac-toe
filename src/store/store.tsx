import { configureStore } from "@reduxjs/toolkit";
import BoardSlicer from "../components/slicers/BoardSlicer";
// import counterSlice from "./slices/counter";

export const store = configureStore({
  reducer: {
    BoardSlicer:BoardSlicer
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;