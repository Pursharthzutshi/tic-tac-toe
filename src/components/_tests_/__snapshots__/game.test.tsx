import {act} from 'react';

/**
 * @jest-environment jsdom
 */

import React from "react";
import { cleanup, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {Board} from "../../Board"
import { checkWinner } from '../../checkWinner';
import { Provider } from 'react-redux';
import { store } from '../../../store/store';

it("snapshot",async ()=>{
    
//   type boardProps = {
//     board: any,
// }

    const {container} = await act(async () => render(
        <Provider store={store}>
        <Board />
        </Provider>
))
    expect(container.querySelector("h3")).toMatchSnapshot();

})


it("checkWinner",async()=>{

  const PlayerWinner = 
  [
    ["X", "O", "O"],
    ["O", "X", "O"],
    ["O", "O", "X"],
  ]


  const OponnentWinner = 
  [
    ["O", "O", "O"],
    ["O", "X", "O"],
    ["O", "O", "X"],
  ]

  expect(checkWinner(OponnentWinner)).toBe("O")
})