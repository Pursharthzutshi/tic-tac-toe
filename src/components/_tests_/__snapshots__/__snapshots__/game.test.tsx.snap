// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<h3>
  Tic Tac Toe Board
</h3>
`;
